### PR TITLE
image/preseed: re-execute snap-preseed --reset

### PR DESCRIPTION
**Pull request by valentindavid**

### CODE REVIEW

File: cmd/snap-preseed/main.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ first-boot startup time`
 
 type options struct {
 	Reset               bool   `long:"reset"`
+	ResetReexeced       bool   `long:"reset-reexeced"`
```
Comment (pedronis) [inline cmd/snap-preseed/main.go:51]: we usually use env var to manage reexec, an option work too, I wonder if it should be marked hidden though, is it is not for direct user consumption. Maybe also we should call it --reset-chroot ?

**Review by pedronis — COMMENTED** [review]
thanks for this, some considerations around the new option

File: image/preseed/preseed_linux.go [hunk 2]
```diff
@@ -443,50 +503,63 @@ func prepareClassicChroot(preseedChroot string) (*targetSnapdInfo, func(), error
...
 	}
 
-	removeMountpoint := func() {
-		if err := os.Remove(where); err != nil {
-			fmt.Fprintf(Stderr, "%v", err)
+	unmountDev := func() {}
+	if !osutil.FileExists("/dev/loop-control") {
+		cmd := exec.Command("mount", "-t", "devtmpfs", "none", "/dev")
+		if out, err := cmd.CombinedOutput(); err != nil {
+			unmountProc()
```
Comment (pedronis) [inline image/preseed/preseed_linux.go:524]: an approach that we use when there is a growing set of undo/cleanup functions to call on error is to have a slice of functions and append to it and then have a common defer that  like:

```
defer func() {
   if err != nil {
       for  i := len(cleanups)-1; i >=0; i-- {
          cleanups[i]()
       }
   }
}()

```

or similar

File: image/preseed/preseed_linux.go [hunk 3]
```diff
@@ -559,6 +632,19 @@ func runPreseedMode(preseedChroot string, targetSnapd *targetSnapdInfo) error {
 	return nil
 }
 
+func runReset(preseedChroot string, targetSnapd *targetSnapdInfo) error {
```
Comment (pedronis) [inline image/preseed/preseed_linux.go:635]: should this be called reexecReset or reexecFoReset ?

**Review by pedronis — COMMENTED** [review]
thank you, some small remarks

File: image/preseed/preseed_linux.go [hunk 4]
```diff
@@ -443,50 +503,64 @@ func prepareClassicChroot(preseedChroot string) (*targetSnapdInfo, func(), error
...
 	}
+	defer func() {
+		cleanup()
+	}()
 
-	// create mountpoint for core/snapd
-	where := filepath.Join(rootDir, snapdMountPath)
-	if err := os.MkdirAll(where, 0755); err != nil {
-		return nil, nil, err
+	if !osutil.FileExists("/proc/self/cmdline") {
```
Comment (pedronis) [inline image/preseed/preseed_linux.go:519]: this if case and the one below don't seem triggered in tests here

**Review by pedronis — APPROVED** [review]
thanks, test question

**Review by mvo5 — APPROVED** [review]
Thank you

**Comment by mvo5** [comment]
Note that I moved the milestone back to 2.59 as it already has a version check in there and 2.59 is fine.

**Pull request merged**